pkg/kubelet/kuberuntime/kuberuntime_container.go

```diff
@@ -611,13 +613,41 @@ func (m *kubeGenericRuntimeManager) killContainer(pod *v1.Pod, containerID kubec
 	return err
 }
 
+func isSidecar(containers []v1.Container, name string) bool {
+	for _, container := range containers {
+		if container.Name == name && lifecycleSidecar(container) {
+			return true
+		}
+	}
+	return false
+}
+
 // killContainersWithSyncResult kills all pod's containers with sync results.
 func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, runningPod kubecontainer.Pod, gracePeriodOverride *int64) (syncResults []*kubecontainer.SyncResult) {
 	containerResults := make(chan *kubecontainer.SyncResult, len(runningPod.Containers))
 	wg := sync.WaitGroup{}
 
-	wg.Add(len(runningPod.Containers))
-	for _, container := range runningPod.Containers {
+	var containers []*kubecontainer.Container
+	var sidecars []*kubecontainer.Container
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarLifecycle) {
+		if pod != nil {
+			for _, container := range runningPod.Containers {
+				if isSidecar(pod.Spec.Containers, container.Name) {
+					sidecars = append(sidecars, container)
+				} else {
+					containers = append(containers, container)
+				}
+			}
+		} else {
+			containers = runningPod.Containers
```

Comment:
we might want to better handle pod is nil case here - if somehow pod is nil, then we are killing all containers together, which is not intended.

either we make sure from caller that pod will never be nil, or we do nothing in this function if pod == nil

Reply:
Pod == nil is a fairly unique edge case, that generally occurs when the kubelet restarts, I think changing this function to do nothing when pod == nil, is quite a disruptive change that changes existing behaviour.
 I think I'd rather caveat the implementation with `shutdown ordering may not be honoured in certain edge cases such as the kubelet restarting during pod termination` then risk breaking the pod termination flow, or having a pod get stuck in terminating.
There are already some caveats with shutdown ordering such as when the node is terminated as the kubelet hands off container termination at this point.

Reply:
Ok if that’d been the case let’s print at least warning logs or even generate events for the fact that we are going to break shutdown ordering so user can debug.

Reply:
yeah that sounds like a good idea